### ISSUE TITLE
How to run wtdbg-1.2.8 in multiple nodes (the usage of --load-xxx)

### ISSUE THREAD

**Issue by iggyB** [ruanjue/wtdbg-1.2.8#7]
Hi Jue,

Very interesting project! I'm eager to test **wtdbg** and interested to make use of several nodes.

You write that this can be achieved by combining **kbm** and **wtdbg --load-alignments**. Could you write a few more words or give an example of commands?

Cheers,
Iggy

**Comment by ruanjue** [comment]
**Case One**: speed up in clusters, but no limit on memory
```
for i in `seq 0 99`; do echo "kbm-1.2.8 -d reads.fa -i <(subsampling_fa.pl reads.fa 100 $i) -t 0 -o reads.kbmap.node$i"; done
```
Here,  subsampling_fa.pl is a peusdo-script that output the $i'th read per 100 input reads.
After all nodes finished the alignments, concatenate all alignments into `reads.kbmap.all`
Then find a big memory PC, run `wtdbg-1.2.8 -i reads.fa --load-alignments reads.kbmap.all -t 0 -o dbg`.

** Case Two **: my nodes have little memory, need to trade time for space
```
for i in `seq 0 99`; do for j in `seq 0 99`; do  echo "kbm-1.2.8 -t 0 -d <(subsampling_fa.pl reads.fa 100 $i) -i <(subsampling_fa.pl reads.fa 100 $j) -o reads.kbmap.$i.$j"; done; done
``` 
After finished all those 100X100 commands, concatenate all, and invoke wtdbg-1.2.8 as in Case One.

Above all, I haven't run it in a cluster by myself. It would be kind of you to share the experiences in future.

Best,
Jue

**Comment by iggyB** [comment]
Thanks for quick response and suggestions. I will be glad to share experience.

**Comment by iggyB** [comment]
Hej Jue,

I'm trying to set up a small test case before moving to actual assembly. Lets say I have 4 SMRT cells.

Scenarious:

**Preprocessing**
1. Select longest subreads per well for all 4 SMRT cells (I know this is done in wtdbg automatically, but I guess it is good to do manually before starting multi node alignments?);

**Single node**
1. concatenate selected longest subreads;
2. run wtbbg pipeline (run_wtdbg_assembly.sh);
3. all good, got assembly.

**Multi node**
1. do 16 separate alignments with kbm-1.2.8 (all SMRT cells against all);
2. concatenate alignments;
3. run wtdbg-1.2.8 with --load-alignments;
4. no assembly, .nodes very small, .lay empty.

A couple of questions:
* any obvious mistake?
* do I need to rename reads in special way (maybe "/" in original names is messing things)?
* do I need to sort reads/alignments in special way?

Cheers,
Iggy

**Comment by ruanjue** [comment]
Hi,

Please give me your command lines:
1, all against all in 16 runs
2, run wtdbg-1.2.8
Also I want to have a look at the log message from wtdbg-1.2.8, maybe too big in log file, I just need the text showed on screen.

REPLY to your  questions:
1, I think all against all had produced duplicated alignments
2, I used to rename sequences using wtdbg-1.2.8/rename_fa.pl or rename_fq.pl, but it just reduce the disk space
3, no, wtdbg will sort them inside

Best,
Jue

**Comment by iggyB** [comment]
Hi,

OK, it's clear then about names and sorting. I think you might be right about duplicate alignments.

Here's the commands:
```
me$: for i in {01..04}; do for j in {01..04}; do echo "kbm-1.2.8 -t 0 -d $i.fasta -i $j.fasta -o reads.kbmap.$i-$j"; done; done
kbm-1.2.8 -t 0 -d 01.fasta -i 01.fasta -o reads.kbmap.01-01
kbm-1.2.8 -t 0 -d 01.fasta -i 02.fasta -o reads.kbmap.01-02
kbm-1.2.8 -t 0 -d 01.fasta -i 03.fasta -o reads.kbmap.01-03
kbm-1.2.8 -t 0 -d 01.fasta -i 04.fasta -o reads.kbmap.01-04
kbm-1.2.8 -t 0 -d 02.fasta -i 01.fasta -o reads.kbmap.02-01
kbm-1.2.8 -t 0 -d 02.fasta -i 02.fasta -o reads.kbmap.02-02
kbm-1.2.8 -t 0 -d 02.fasta -i 03.fasta -o reads.kbmap.02-03
kbm-1.2.8 -t 0 -d 02.fasta -i 04.fasta -o reads.kbmap.02-04
kbm-1.2.8 -t 0 -d 03.fasta -i 01.fasta -o reads.kbmap.03-01
kbm-1.2.8 -t 0 -d 03.fasta -i 02.fasta -o reads.kbmap.03-02
kbm-1.2.8 -t 0 -d 03.fasta -i 03.fasta -o reads.kbmap.03-03
kbm-1.2.8 -t 0 -d 03.fasta -i 04.fasta -o reads.kbmap.03-04
kbm-1.2.8 -t 0 -d 04.fasta -i 01.fasta -o reads.kbmap.04-01
kbm-1.2.8 -t 0 -d 04.fasta -i 02.fasta -o reads.kbmap.04-02
kbm-1.2.8 -t 0 -d 04.fasta -i 03.fasta -o reads.kbmap.04-03
kbm-1.2.8 -t 0 -d 04.fasta -i 04.fasta -o reads.kbmap.04-04

me$: cat reads.kbmap.0* > reads.kbmap.all
me$: cat *.fasta > reads.fasta
me$: wtdbg-1.2.8 -t 0 -i reads.fasta --load-alignments reads.kbmap.all -fo dbg
```

wtdbg log from successful **single node** run:
```
[Wed May 23 08:55:45 2018] loading reads 
1132070 reads
[Wed May 23 08:56:39 2018] Done, 1132070 reads, 13339262486 bp, 51541324 bins
[Wed May 23 08:56:39 2018] generating nodes, 6 threads
[Wed May 23 08:56:39 2018] indexing bins[0,51541324] (13194578944 bp), 6 threads
[Wed May 23 08:56:39 2018] - scanning kmers (K0P21 subsampling 1/4) from 51541324 bins
51541324 bins
[Wed May 23 09:00:59 2018] - Total kmers = 385220189
[Wed May 23 09:00:59 2018] - average kmer depth = 4
[Wed May 23 09:00:59 2018] - 421729358 low frequency kmers (<2)
[Wed May 23 09:00:59 2018] - 5899 high frequency kmers (>1000)
[Wed May 23 09:00:59 2018] - indexing 385214290 kmers, 1568251421 instances 
51541324 bins
[Wed May 23 09:06:44 2018] - masked 72284 bins as closed
[Wed May 23 09:06:44 2018] - sorting
[Wed May 23 09:06:48 2018] Done
1132070 reads|total hits 2808747
[Wed May 23 09:19:03 2018] chainning ...  93853 hits into 46583
[Wed May 23 09:19:03 2018] picking best 500 hits for each read ... 2128453 hits
[Wed May 23 09:19:04 2018] clipping ... 97.01% bases
[Wed May 23 09:19:06 2018] generated 14617181 regs
[Wed May 23 09:19:06 2018] sorting regs ...  Done
[Wed May 23 09:19:07 2018] generating intervals ...  1065962 intervals
[Wed May 23 09:19:07 2018] selecting important intervals from 1065962 intervals
[Wed May 23 09:19:18 2018] Intervals: kept 48173, discarded 1017789
[Wed May 23 09:19:18 2018] Done, 48173 nodes
[Wed May 23 09:19:18 2018] output "dbg.1.nodes". Done.
[Wed May 23 09:19:18 2018] median node depth = 4
[Wed May 23 09:19:18 2018] masked 21 high coverage nodes (>200 or <3)
[Wed May 23 09:19:18 2018] masked 1152 repeat-like nodes by local subgraph analysis
[Wed May 23 09:19:18 2018] generating edges
[Wed May 23 09:19:18 2018] Done, 113663 edges
[Wed May 23 09:19:18 2018] output "dbg.1.reads". Done.
[Wed May 23 09:19:19 2018] output "dbg.1.dot". Done.
[Wed May 23 09:19:19 2018] graph clean
[Wed May 23 09:19:19 2018] rescued 718 low cov edges
[Wed May 23 09:19:19 2018] deleted 0 binary edges
[Wed May 23 09:19:19 2018] deleted 38372 isolated nodes
[Wed May 23 09:19:19 2018] cut 322 transitive edges
[Wed May 23 09:19:19 2018] output "dbg.2.dot". Done.
[Wed May 23 09:19:19 2018] 46 bubbles; 511 tips; 1 yarns;
[Wed May 23 09:19:19 2018] deleted 42 isolated nodes
[Wed May 23 09:19:19 2018] output "dbg.3.dot". Done.
[Wed May 23 09:19:19 2018] cut 4 branching nodes
[Wed May 23 09:19:19 2018] deleted 1 isolated nodes
[Wed May 23 09:19:19 2018] building unitigs
[Wed May 23 09:19:19 2018] TOT 9888256, CNT 2129, AVG 4645, MAX 68096, N50 5888, L50 419, N90 2048, L90 1631, Min 1280
[Wed May 23 09:19:19 2018] output "dbg.frg.nodes". Done.
[Wed May 23 09:19:19 2018] generating links
[Wed May 23 09:19:19 2018] generated 737 links
[Wed May 23 09:19:19 2018] output "dbg.frg.dot". Done.
[Wed May 23 09:19:19 2018] rescue 0 weak links
[Wed May 23 09:19:19 2018] deleted 56 binary links
[Wed May 23 09:19:19 2018] cut 13 transitive links
[Wed May 23 09:19:19 2018] remove 108 boomerangs
[Wed May 23 09:19:19 2018] detached 4 repeat-associated paths
[Wed May 23 09:19:19 2018] remove 167 weak branches
[Wed May 23 09:19:19 2018] cut 1 tips
[Wed May 23 09:19:19 2018] pop 0 bubbles
[Wed May 23 09:19:19 2018] cut 0 tips
[Wed May 23 09:19:19 2018] output "dbg.ctg.dot". Done.
[Wed May 23 09:19:19 2018] building contigs
[Wed May 23 09:19:19 2018] searched 1842 contigs
[Wed May 23 09:19:19 2018] Estimated: TOT 7223552, CNT 567, AVG 12740, MAX 62464, N50 14848, L50 153, N90 6400, L90 438, Min 5120
[Wed May 23 09:19:19 2018] output 567 contigs
[Wed May 23 09:19:19 2018] Program Done
```

wtdbg log from unsuccessful **multi node** run:
```
[Fri May 25 15:33:02 2018] loading reads
1132070 reads
[Fri May 25 15:33:54 2018] Done, 1132070 reads, 13339262486 bp, 51541324 bins
[Fri May 25 15:33:54 2018] loading alignments from "reads.kbmap.all"
11857517 lines, 11857517 hits
[Fri May 25 15:34:12 2018] chainning ...  280114 hits into 139031
[Fri May 25 15:34:13 2018] picking best 500 hits for each read ... 6176280 hits
[Fri May 25 15:34:14 2018] clipping ... 1.09% bases
[Fri May 25 15:34:22 2018] generated 60379375 regs
[Fri May 25 15:34:23 2018] sorting regs ...  Done
[Fri May 25 15:34:25 2018] generating intervals ...  12462329 intervals
[Fri May 25 15:34:26 2018] selecting important intervals from 12462329 intervals
[Fri May 25 15:36:29 2018] Intervals: kept 1, discarded 12462328
[Fri May 25 15:36:29 2018] Done, 1 nodes
[Fri May 25 15:36:29 2018] output "dbg.1.nodes". Done.
[Fri May 25 15:36:29 2018] median node depth = 0
[Fri May 25 15:36:29 2018] masked 0 high coverage nodes (>200 or <3)
[Fri May 25 15:36:29 2018] masked 0 repeat-like nodes by local subgraph analysis
[Fri May 25 15:36:29 2018] generating edges
[Fri May 25 15:36:29 2018] Done, 1 edges
[Fri May 25 15:36:29 2018] output "dbg.1.reads". Done.
[Fri May 25 15:36:30 2018] output "dbg.1.dot". Done.
[Fri May 25 15:36:30 2018] graph clean
[Fri May 25 15:36:30 2018] deleted 0 binary edges
[Fri May 25 15:36:30 2018] deleted 1 isolated nodes
[Fri May 25 15:36:30 2018] cut 0 transitive edges
[Fri May 25 15:36:30 2018] output "dbg.2.dot". Done.
[Fri May 25 15:36:30 2018] deleted 0 isolated nodes
[Fri May 25 15:36:30 2018] output "dbg.3.dot". Done.
[Fri May 25 15:36:30 2018] cut 0 branching nodes
[Fri May 25 15:36:30 2018] deleted 0 isolated nodes
[Fri May 25 15:36:30 2018] building unitigs
[Fri May 25 15:36:30 2018] 
[Fri May 25 15:36:30 2018] output "dbg.frg.nodes". Done.
[Fri May 25 15:36:30 2018] generating links
[Fri May 25 15:36:30 2018] generated 1 links
[Fri May 25 15:36:30 2018] output "dbg.frg.dot". Done.
[Fri May 25 15:36:30 2018] rescue 0 weak links
[Fri May 25 15:36:30 2018] deleted 0 binary links
[Fri May 25 15:36:30 2018] cut 0 transitive links
[Fri May 25 15:36:30 2018] remove 0 boomerangs
[Fri May 25 15:36:30 2018] detached 0 repeat-associated paths
[Fri May 25 15:36:30 2018] remove 0 weak branches
[Fri May 25 15:36:30 2018] cut 0 tips
[Fri May 25 15:36:30 2018] pop 0 bubbles
[Fri May 25 15:36:30 2018] cut 0 tips
[Fri May 25 15:36:30 2018] output "dbg.ctg.dot". Done.
[Fri May 25 15:36:30 2018] building contigs
[Fri May 25 15:36:30 2018] searched 0 contigs
[Fri May 25 15:36:30 2018] Estimated: 
[Fri May 25 15:36:30 2018] output 0 contigs
[Fri May 25 15:36:30 2018] Program Done
```

Cheers,
Iggy

**Comment by ruanjue** [comment]
The main difference between the alignments in single-node and multi-node is self-alignments in multi-nodes. Another issue, to make the multi-node working as single-node, the only way is:
[single-node]  add --aln-noskip to wtdbg-1.2.8
[multi-node] add -c  to kbm-1.2.8
Because the ability to find contained reads are different when providing part or full reads as reference.

Best,
Jue

**Comment by iggyB** [comment]
Hi Jue,

I tried your suggestions, but still have no success in getting similar results from **single node** versus **multi node** scenarios. 

I really would like to find a solution, so perhaps you will have some more ideas :)

To make it even more simple, I used just 3 smrt cells in the new test:
```
01.fasta
02.fasta
03.fasta
me$: cat *.fasta > 3smrts.fasta
```
Then I run **single node** option
```
me$: wtdbg-1.2.8 -t 6 -i 3smrts.fasta -fo 3smrts --rescue-low-cov-edges --aln-noskip
me$: wtdbg-cns -t 6 -i 3smrts.ctg.lay -fo 3smrts.ctg.lay.fa -c 0
...
Estimated: TOT 4868608, CNT 388, AVG 12548, MAX 89344, N50 14592, L50 105, N90 6656, L90 302, Min 5120
...
```
OK, **~5Mb** total, **~90kb** max.

Then I run **multi node** option:
```
me$: kbm-1.2.8 -d 01.fasta -i 01.fasta -t 12 -C -o reads.kbmap.01-01
me$: kbm-1.2.8 -d 01.fasta -i 02.fasta -t 12 -C -o reads.kbmap.01-02
me$: kbm-1.2.8 -d 01.fasta -i 03.fasta -t 12 -C -o reads.kbmap.01-03
me$: kbm-1.2.8 -d 02.fasta -i 01.fasta -t 12 -C -o reads.kbmap.02-01
me$: kbm-1.2.8 -d 02.fasta -i 02.fasta -t 12 -C -o reads.kbmap.02-02
me$: kbm-1.2.8 -d 02.fasta -i 03.fasta -t 12 -C -o reads.kbmap.02-03
me$: kbm-1.2.8 -d 03.fasta -i 01.fasta -t 12 -C -o reads.kbmap.03-01
me$: kbm-1.2.8 -d 03.fasta -i 02.fasta -t 12 -C -o reads.kbmap.03-02
me$: kbm-1.2.8 -d 03.fasta -i 03.fasta -t 12 -C -o reads.kbmap.03-03
me$: cat reads.kbmap.0* | awk '$1 != $6' > reads.kbmap.all
me$: wtdbg-1.2.8 -t 6 -i 3smrts.fasta --load-alignments reads.kbmap.all -fo multi3smrts --rescue-low-cov-edges
me$: wtdbg-cns -t 6 -i multi3smrts.ctg.lay -fo multi3smrts.ctg.lay.fa -c 0
...
Estimated: TOT 819712, CNT 94, AVG 8721, MAX 27648, N50 7936, L50 29, N90 5376, L90 79, Min 5120
...
```
Hmm, **800kb** total, **27kb** max...

Then I was wondering if it is necessary to have all reads as reference or is it enough just to compare them all?
So I did following:
```
me$: cat reads.kbmap.01-01 reads.kbmap.01-02 reads.kbmap.01-03 reads.kbmap.02-02 reads.kbmap.02-03 reads.kbmap.03-03 | awk '$1 != $6' > reads.kbmap.partly
me$: wtdbg-1.2.8 -t 6 -i 3smrts.fasta --load-alignments reads.kbmap.partly -fo partly3smrts --rescue-low-cov-edges
me$: wtdbg-cns -t 6 -i partly3smrts.ctg.lay -fo partly3smrts.ctg.lay.fa -c 0
...
Estimated: TOT 3699968, CNT 349, AVG 10602, MAX 45312, N50 11776, L50 103, N90 5888, L90 280, Min 5120
...
```
OK, so **3.6Mb** total, **45kb** max. Not sure how to interpret this.

And lastly I did on more test using all reads as reference and mapping single cells like this:
```
me$: kbm-1.2.8 -d 3smrts.fasta -i 01.fasta -t 12 -o reads.kbmap.3smrts-01
me$: kbm-1.2.8 -d 3smrts.fasta -i 02.fasta -t 12 -o reads.kbmap.3smrts-02
me$: kbm-1.2.8 -d 3smrts.fasta -i 03.fasta -t 12 -o reads.kbmap.3smrts-03
me$: cat reads.kbmap.3smrts-0* | awk '$1 != $6' > reads.kbmap.3align
me$: wtdbg-1.2.8 -t 6 -i 3smrts.fasta --load-alignments reads.kbmap.3align -fo 3align3smrts --rescue-low-cov-edges
me$: wtdbg-cns -t 6 -i 3align3smrts.ctg.lay -fo 3align3smrts.ctg.lay.fa -c 0
...
Estimated: TOT 435968, CNT 55, AVG 7927, MAX 22272, N50 7680, L50 20, N90 5376, L90 47, Min 5120
...
```
Hmm, **400kb** total, **22kb** max...

So I wonder if it is possible in general to get **multi node** option to work?

Any thoughts, ideas or discussion is very welcome.

Cheers,
Iggy

**Comment by ruanjue** [comment]
I am afraid you used wrong parameters for kbm, should be `-c` not `-C`. However, I will test the yeast pacbio dataset, and find a solution for you.

**Comment by iggyB** [comment]
hmm
```
kbm-1.2.8: invalid option -- 'c'
```
:)

**Comment by ruanjue** [comment]
```
> kbm-1.2.8 -h
Program: kbm is a simple instance which implemented kmer-binmap
         it maps query sequence against reference by kmer matching
         matched kmer-pairs are bined (256bp) and counted in a matrix
         dynamic programming is used to search the best path
Version: 1.2.8 r1 (20171129)
```
The same version?

**Comment by iggyB** [comment]
```
Version: 1.2.8 (20170830)
```

apparently not...

kbm.c on git has:
```
	fprintf(stdout, "Version: 1.2.8 (20170830)\n");
```

**Comment by ruanjue** [comment]
Sorry, I forgot to update the invoker kbm.c for so many months. Now, it is updated.

**Comment by iggyB** [comment]
Great! No worries.

Could you still do your own test with PacBio yeast dataset and see if you can find a way to get consistent results?

And could you comment on the need to have all reads as reference? Is it enough to compare them all?

Cheers,
Iggy

**Comment by ruanjue** [comment]
The problem was located at the duplicate alignments, e.g. A -> B, and B -> A should only keep one. Now, I modified wtdbg.c to automaticly deal with those cases. 

My test on yeast pacbio dataset `rds.fa`
```
$> split_seqs_2.pl 3 rds.fa
$> echo run single node
$> wtdbg-1.2.8 -t 32 -i rds.fa -fo asm0 --aln-noskip --rescue-low-cov-edges >log.0 2>&1
[Tue May 29 15:42:08 2018] Estimated: TOT 12485376, CNT 49, AVG 254804, MAX 987648, N50 691456, L50 8, N90 252416, L90 18, Min 5120
$> echo run multi nodes
$> seq 1 3 | xargs -i echo kbm-1.2.8 -t 16 -d rds.fa -i rds.fa.shuffle{} -c -fo {}.kbmap \>log.{} 2\>\&1 | sh
$> wtdbg-1.2.8 -t 16 -i rds.fa --load-alignments 1.kbmap --load-alignments 2.kbmap --load-alignments 3.kbmap -fo dbg --rescue-low-cov-edges > log.4 2>&1
[Tue May 29 16:27:01 2018] Estimated: TOT 12507648, CNT 50, AVG 250153, MAX 1303552, N50 558080, L50 8, N90 192000, L90 21, Min 5120
```
It showed that wtdbg can assemble  from kbm's alignments.

Best,
Jue

**Comment by ruanjue** [comment]
By add **--no-read-length-sort**, you can easily avoid much of duplicated alignments:
```
-d 01 -i 01
-d 01 -i 02
-d 01 -i 03
-d 02 -i 02
-d 02 -i 03
-d 03 -i 03
```

To speed it up, I suggest to test about without `-c` in kbm, and have a look at the assembly metrics.
